internal/releasesources/release_sources.go

```diff
@@ -0,0 +1,510 @@
+				if shouldSwitch {
+					cpuArchitecture = apiCpuArchitecture
+				}
+
+				graph, err := h.releasesClient.getReleases(channel, openshiftVersion, cpuArchitecture)
+				if err != nil {
+					h.log.WithError(err).Warnf("failed to get releases from api with: %s-%s, %s", channel, openshiftVersion, cpuArchitecture)
+					continue
+				}
+
+				if shouldSwitch {
+					if arch, ok := cpuArchMapFromAPIArch[cpuArchitecture]; ok {
+						cpuArchitecture = arch
+					}
+				}
+
+				for _, node := range graph.Nodes {
+					majorMinorVersion, err := common.GetMajorMinorVersion(node.Version)
+					if err != nil {
+						h.log.WithError(err).Warnf("failed to get the major.minor version of: %s", node.Version)
+						continue
+					}
+
+					if openshiftVersion == *majorMinorVersion {
+						// Got a match for OpenShift ReleaseImage
+						newReleaseImage, err := h.createReleaseImage(channel, node.Version, cpuArchitecture)
+						if err != nil {
+							return nil, err
+						}
+						releases = append(releases, newReleaseImage)
+					}
+				}
+			}
+		}
+	}
+
+	h.log.Debugf("Got %d releases", len(releases))
+	return releases, nil
+}
+
+func (h *releaseSourcesHandler) getStaticReleaseImages() []*common.ReleaseImage {
+	dbReleaseImages := []*common.ReleaseImage{}
+	for _, releaseImage := range h.staticReleaseImages {
+		dbReleaseImage := common.ReleaseImage{ReleaseImage: *releaseImage}
+		// For backward compatibility with release images that lack the CPUArchitectures field.
+		if dbReleaseImage.CPUArchitectures == nil {
+			dbReleaseImage.CPUArchitectures = []string{*dbReleaseImage.CPUArchitecture}
+		}
+		dbReleaseImage.DBCPUArchitectures = dbReleaseImage.CPUArchitectures
+		dbReleaseImages = append(dbReleaseImages, &dbReleaseImage)
+	}
+
+	return dbReleaseImages
+}
+
+// mergeDBReleaseImages merges static and dynamic release images together such that
+// static release images have precedence in case of conflicts e.g. which release is default
+func (h *releaseSourcesHandler) mergeDBReleaseImages(staticReleaseImages, dynamicReleaseImages []*common.ReleaseImage) []*common.ReleaseImage {
+	mergedReleaseImages := []*common.ReleaseImage{}
+	staticReleaseImagesReferenceSet := map[string]bool{}
+	defaultStaticReleaseExists := false
+
+	for _, staticRelease := range staticReleaseImages {
+		mergedReleaseImages = append(mergedReleaseImages, staticRelease)
+		staticReleaseImagesReferenceSet[*staticRelease.URL] = true
+		if staticRelease.Default {
+			defaultStaticReleaseExists = true
+		}
+	}
+
+	for _, dynamicRelease := range dynamicReleaseImages {
+		if staticReleaseImagesReferenceSet[*dynamicRelease.URL] {
+			continue
+		}
+		if dynamicRelease.Default && defaultStaticReleaseExists {
+			dynamicRelease.Default = false
+		}
+		mergedReleaseImages = append(mergedReleaseImages, dynamicRelease)
+	}
+
+	return mergedReleaseImages
+}
+
+func (h *releaseSourcesHandler) syncReleaseImagesWithErr() error {
```

Comment:
I think it should be blocked with leader elector no? What will happened if 2 instances will run that code at the same time?

Reply:
Every run all the previous releases are replaced with new ones inside a transaction, so we should be good in that context. Will be more efficient though to have only one instance running it  instead of all